### PR TITLE
Casting low-rank CholeskyPivoted to Matrix

### LINKED ISSUE #680: Converting a low rank CholeskyPivoted with Matrix() does not return the correct result
The current implementation of `Matrix` on a CholeskyPivoted object returns the wrong result if the factorization is of low rank.

```julia
A = randn(3,1)
B = A*A'
C = cholesky(B, Val(true), tol = 1e-8, check=false)
```
Then B != Matrix(C). 

If we instead defined
```julia
function Matrix(C::CholeskyPivoted)
    ip = invperm(C.p)
    U = C.U[1:C.rank, ip]
    U'U
end
```
We would get the correct result. Other functions, like `det` already implement the correct behavior in this case, by using the `rank` field. Therefore, I think this change would make the behavior consistent for low rank factorizations.

### LINKED ISSUE #33704: Converting a low rank CholeskyPivoted with Matrix() does not return the correct result
The current implementation of `Matrix` on a CholeskyPivoted object returns the wrong result if the factorization is of low rank.

```julia
A = randn(3,1)
B = A*A'
C = cholesky(B, Val(true), tol = 1e-8, check=false)
```
Then B != Matrix(C). 

If we instead defined
```julia
function Matrix(C::CholeskyPivoted)
    ip = invperm(C.p)
    U = C.U[1:C.rank, ip]
    U'U
end
```
We would get the correct result. Other functions, like `det` already implement the correct behavior in this case, by using the `rank` field. Therefore, I think this change would make the behavior consistent for low rank factorizations.

### PR DESCRIPTION
**Pull request by SebastianAment**
I created and closed a pull request for this already, because I accidentally force pushed to my branch during the time the PR was open. I apologize for the duplicate. 

Fixes JuliaLang/LinearAlgebra.jl#680.

### CODE REVIEW

File: stdlib/LinearAlgebra/test/cholesky.jl [hunk 1]
```diff
@@ -392,4 +392,14 @@ end
...
 end
 
+@testset "issue #33704, casting low-rank CholeskyPivoted to Matrix" begin
+    T = Float64
+    n = 8
+    tol = 1e-8
+    A = randn(T, (n,1))
+    B = A*A'
+    C = cholesky(B, Val(true), tol = tol, check=false)
+    @test maximum(abs, B-Matrix(C)) < tol
```
Comment (andreasnoack) [inline stdlib/LinearAlgebra/test/cholesky.jl]:
```suggestion
    @test B ≈ Matrix(C)
```
and then you can delete the `tol` definition above.

**Review by andreasnoack — COMMENTED** [review]
Thanks for the PR. Looks good. Just a minor suggestion in the test part.

File: stdlib/LinearAlgebra/test/cholesky.jl (hunk 1 above)
Comment (SebastianAment) [reply stdlib/LinearAlgebra/test/cholesky.jl]: Great, code is updated now.

**Review by andreasnoack — APPROVED** [review]

File: stdlib/LinearAlgebra/test/cholesky.jl [hunk 2]
```diff
@@ -392,4 +392,13 @@ end
 
 end
 
+@testset "issue #33704, casting low-rank CholeskyPivoted to Matrix" begin
+    T = Float64
+    n = 8
+    A = randn(T, (n,1))
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/test/cholesky.jl]: `randn(8,1)` will return would you're asking here for, and neither `T` nor `n` are used again. So why not reduce the number of lines here?

**Review by dkarrasch — COMMENTED** [review]
LGTM, except for a minor test simplification suggestion.

File: stdlib/LinearAlgebra/test/cholesky.jl (hunk 2 above)
Comment (SebastianAment) [reply stdlib/LinearAlgebra/test/cholesky.jl]: I made the change. Let me know if there is anything else.

**Pull request merged**